#Finding the most accurate classification algorithm from


1.   Logisitic Regression
2.   K-Nearest Neighbhour
3.    Naive Bayes
4.    Decision Tree (ID3)
5.    Support Vector M



In [22]:
import numpy as np
import pandas as pd
from sklearn.preprocessing import StandardScaler
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score,confusion_matrix,classification_report,f1_score,recall_score

In [12]:
df=pd.read_csv("/content/heart_cleaned.csv")
df.head()

,Unnamed: 0,Age,is_Male,RestingBP,Cholesterol,FastingBS,MaxHR,is_ExerciseAngina,Oldpeak,HeartDisease,ChestPainType_ATA,ChestPainType_NAP,RestingECG_Normal,RestingECG_ST,ST_Slope_Flat,ST_Slope_Up
0,0,40,1,140.0,289.0,0,172,0,0.0,0,1,0,1,0,0,1
1,1,49,0,160.0,180.0,0,156,0,1.0,1,0,1,1,0,1,0
2,2,37,1,130.0,283.0,0,98,0,0.0,0,1,0,0,1,0,1
3,3,48,0,138.0,214.0,0,108,1,1.5,1,0,0,1,0,1,0
4,4,54,1,150.0,195.0,0,122,0,0.0,0,0,1,1,0,0,1


In [13]:
df.columns

Index(['Unnamed: 0', 'Age', 'is_Male', 'RestingBP', 'Cholesterol', 'FastingBS',
       'MaxHR', 'is_ExerciseAngina', 'Oldpeak', 'HeartDisease',
       'ChestPainType_ATA', 'ChestPainType_NAP', 'RestingECG_Normal',
       'RestingECG_ST', 'ST_Slope_Flat', 'ST_Slope_Up'],
      dtype='object')

In [14]:
df=df.drop(columns=['Unnamed: 0'])

In [15]:
df.head(3)

,Age,is_Male,RestingBP,Cholesterol,FastingBS,MaxHR,is_ExerciseAngina,Oldpeak,HeartDisease,ChestPainType_ATA,ChestPainType_NAP,RestingECG_Normal,RestingECG_ST,ST_Slope_Flat,ST_Slope_Up
0,40,1,140.0,289.0,0,172,0,0.0,0,1,0,1,0,0,1
1,49,0,160.0,180.0,0,156,0,1.0,1,0,1,1,0,1,0
2,37,1,130.0,283.0,0,98,0,0.0,0,1,0,0,1,0,1


In [16]:
X=df.drop("HeartDisease",axis=1)
y=df["HeartDisease"]

X_train, X_test, y_train, y_test = train_test_split(
   X, y, test_size=0.20, random_state=42)     #Test split 20

In [17]:
scaler=StandardScaler()

In [18]:
X_train_scaled=scaler.fit_transform(X_train)
X_test_scaled=scaler.fit_transform(X_test)

In [19]:
from sklearn.linear_model import LogisticRegression
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC


In [29]:
models={
    "Logistic_Reg":LogisticRegression(max_iter=10000),
    "K_NN":KNeighborsClassifier(),
    "NB": GaussianNB(),
    "DT":DecisionTreeClassifier(),
    "SVM":SVC()
}

In [30]:
results=[]
results_s=[]

In [31]:
for name,model in models.items():
  model.fit(X_train,y_train)
  y_pred=model.predict(X_test)
  acc=accuracy_score(y_test,y_pred)
  f1=f1_score(y_test,y_pred)
  re=recall_score(y_test,y_pred)
  results.append({
      "model":name,
      "Accuracy":acc,
      "F1":f1,
      "Re":re
  })

In [33]:
for name,model in models.items():
  model.fit(X_train_scaled,y_train)
  y_pred=model.predict(X_test_scaled)
  acc=accuracy_score(y_test,y_pred)
  f1=f1_score(y_test,y_pred)
  re=recall_score(y_test,y_pred)
  results_s.append({
      "model":name,
      "Accuracy":acc,
      "F1":f1,
      "Re":re
  })

In [34]:
print("Not Scaled",results)
print("\n\n\n")
print("Scaled",results_s)

Not Scaled [{'model': 'Logistic_Reg', 'Accuracy': 0.8478260869565217, 'F1': 0.8653846153846154, 'Re': 0.8411214953271028}, {'model': 'K_NN', 'Accuracy': 0.6739130434782609, 'F1': 0.7169811320754716, 'Re': 0.7102803738317757}, {'model': 'NB', 'Accuracy': 0.842391304347826, 'F1': 0.8599033816425121, 'Re': 0.8317757009345794}, {'model': 'DT', 'Accuracy': 0.7934782608695652, 'F1': 0.8118811881188119, 'Re': 0.7663551401869159}, {'model': 'SVM', 'Accuracy': 0.6630434782608695, 'F1': 0.696078431372549, 'Re': 0.6635514018691588}]




Scaled [{'model': 'Logistic_Reg', 'Accuracy': 0.8532608695652174, 'F1': 0.8720379146919431, 'Re': 0.8598130841121495}, {'model': 'K_NN', 'Accuracy': 0.8478260869565217, 'F1': 0.8679245283018868, 'Re': 0.8598130841121495}, {'model': 'NB', 'Accuracy': 0.842391304347826, 'F1': 0.8599033816425121, 'Re': 0.8317757009345794}, {'model': 'DT', 'Accuracy': 0.782608695652174, 'F1': 0.8095238095238095, 'Re': 0.794392523364486}, {'model': 'SVM', 'Accuracy': 0.842391304347826,

#Scaled Logistic Regression is the best as
Accuracy: 85.3%
F1 Score: 87.2%
Recall: 86.0%
#Runner-Up is Scaled KNN
Accuracy: 84.8%
F1: 86.8%


Clear second place. Better than SVM on all metrics except Recall.

In [35]:
import joblib
joblib.dump(models["Logistic_Reg"],"logistic_heart.pkl")
joblib.dump(scaler,"scaler.pkl")
joblib.dump(X.columns.tolist(),"columns.pkl")

['columns.pkl']